### DISCUSSION TITLE
How to deal with a csv file from a html page in a fetch() POST request.

### DISCUSSION THREAD

**Discussion by Inrap** [PostgREST/postgrest#2914] [Q&A] [answered]
_beginner with JS_

The goal is to select a CSV file on a computer to insert the data in a distant PostgreSQL table using `POST` request.

HTML could look like this :

    <!DOCTYPE html >
    <html lang="fr">
      <head>
         <meta chardet="UTF-8" >
         <link  rel="stylesheet" href="./style.css" >
         <title>postgrest</title>
      </head>
      <body>
        <div class="fichier-csv-container">
            <label for="fichier-csv" ></label></label><input type="file" class="fichier-csv" id="fichier-csv" name="fichiercsv" required></input>
            <p class="btn-integrer">integrer</p>
        </div>
      </body>
     <script type="module" src="./testpost.js" ></script>
  </html> 

Once the CSV file has been selected, a click on `btn-integrer` causes it to be processed in the browser.

    const btnIntegrer = document.querySelector(".btn-integrer")

    btnIntegrer.addEventListener("click", () => {

    const fichierCsvInput = document.querySelector(".fichier-csv")
    const fichierCsv = fichierCsvInput.files[0]
    reader.readAsText(fichierCsv)
    let data = ''
    let datafinal = ``
    reader.addEventListener("load", (e) => {
        data = e.target.result
        const ligneTab = data.split("\r\n")
        ligneTab.forEach( ligne => datafinal += `${ligne}\n` )
    })

    fetch('http://blabla/testapi?', {
        method: "POST",
        headers: {
            "Content-Type": "text/csv"
        },
        body: datafinal
        })
        .then(result => {
            console.log('insertion réussie:', result)
        } )
        .catch(err => {
            console.assert.error('insertion échouée:', err)
        })
      })

I don't know how to deal with the CSV file to get a correct `body` in the `load event` like in the [doc](https://postgrest.org/en/v10.2/api.html#insertions).
I have an error in the console : 
`code	"PGRST102"
details	null
hint	null
message	'parse error (not enough input) at ""'`
Obviously, there is an issue with `datafinal` but I can't get rid of it.

*EDIT*
my original csv : 
num,description
3,bli bli
4,bli

Thanks

**Comment by Inrap** [comment, marked as answer]
the problem was that `fetch` is async so `datafinal` was empty. I moved the `fetch` inside the `load` event. Now it is good.

**Reply by laurenceisla** [reply]
Cool, so this solved your issue? You can mark your own answer as "answered" if that's so.
